feat: Add client side metrics collection to readRows calls

danieljbruce opened the pull request:
**Summary:**

This PR will collect client side metrics for readRows calls.

**Background:**

The following PRs have been added to support the plumbing for client-side metrics.
https://github.com/googleapis/nodejs-bigtable/pull/1569
https://github.com/googleapis/nodejs-bigtable/pull/1566

They didn't change the client behaviour at all. This PR starts using this plumbing for readRows calls.

The diagram below shows the CSM architecture with new pieces in green in the diagram:

![CSM With fixture - 3rd PR - ReadRows calls (1)](https://github.com/user-attachments/assets/7f1207e4-55c2-4bb0-8dac-796bc003720a)

**Changes:**

`src/client-side-metrics/operation-metrics-collector.ts:` Added optimizations and added guards to make sure that if the data from the server is of some unexpected form that it won't crash our clients.
`src/index.ts:` Add code to generate a uuid for each client and store the GCP metrics handler on the client. Also add piping to pass status events along and add a setting that lets users turn off client side metrics.
`src/tabular-api-surface.ts`: Use the metrics collector with the gcp metrics handler to collect client side metrics in readRows calls
`system-test/client-side-metrics-setup-table.ts:` A test utility function for setting up an instance with a few tables
`system-test/client-side-metrics.ts:` A file with some tests for capturing various scenarios where a user might be collecting client side metrics that ensures the right metrics are exported without error even when multiple tables, instances or clients are involved.
`system-test/read-rows.ts:` This test has a mock that is client side metrics incompatible so don't use client side metrics in this test.

**Next Steps:**

- Add plumbing for application blocking latencies
- Collect client side metrics for other grpc endpoints

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into 359913994-exporter-PR

# Conflicts:
#	package.json
#	src/client-side-metrics/client-side-metrics-attributes.ts
#	src/client-side-metrics/metrics-handler.ts
#	test/metrics-collector/typical-method-call.txt

danieljbruce committed:
Increase the timeout

danieljbruce committed:
Use the PushMetricExporter interface

danieljbruce committed:
Removed some interfaces that are not used anymore

danieljbruce committed:
Update JSdoc

danieljbruce committed:
Move fake start time and fake end time

danieljbruce committed:
Remove the TODO

danieljbruce committed:
Update documentation

danieljbruce committed:
Add additional information to the error reported

danieljbruce committed:
Move start time and end time

danieljbruce committed:
Try to use timestamps in order

danieljbruce committed:
Reduce timestamp delay

danieljbruce committed:
Remove only

danieljbruce committed:
Inline addFakeRecentTimestamps

danieljbruce committed:
Move replace timestamps into the only file it

is used

danieljbruce committed:
Fix comment

danieljbruce committed:
Rename the metric types

danieljbruce committed:
Generate documentation for the new classes

danieljbruce committed:
Add documentation for monitored resource

danieljbruce committed:
Generate documentation for the other metrics

danieljbruce committed:
Generate documentation for the constructor

danieljbruce committed:
Get documentation for replaceTimestamps, fixtures

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into 359913994-exporter-PR

danieljbruce committed:
Reduce the interval time

danieljbruce committed:
Make view list based on latencies

danieljbruce committed:
Add a guard for count

danieljbruce committed:
Add a CSM test for Bigtable

danieljbruce committed:
Add handlers to the middleware layer

danieljbruce committed:
Add the metrics to the tabular api surface

danieljbruce committed:
Move the tests

danieljbruce committed:
Use the GCPMetricsHandler in the tabular api surface

danieljbruce committed:
Add a guard in the metrics collector

danieljbruce committed:
Fix the proxyquire issue

danieljbruce committed:
Export every 60 seconds instead

danieljbruce committed:
Add a guard to the metadata collector

danieljbruce committed:
Add a guard to the export function

danieljbruce committed:
Make sure to await to avoid a race condition

danieljbruce committed:
Adjust headers

danieljbruce committed:
Add the right assertions for the TestMetricsHandle

danieljbruce committed:
Send the project id to the mock

danieljbruce committed:
Invert the for and if

danieljbruce committed:
Pull all attributes out

danieljbruce committed:
Eliminate the need for the type guard

danieljbruce committed:
Eliminate the data points interface

danieljbruce committed:
only

danieljbruce committed:
Revert "Eliminate the need for the type guard"

This reverts commit 791e70dcb71cbf067eac351da6f62ef438be3d5c.

danieljbruce committed:
Pull timeseries into one variable

Solve the descriptor issue

danieljbruce committed:
Eliminate an unused import

danieljbruce committed:
Add a comment that explains the usefulness

of each metric attribute category

danieljbruce committed:
Remove the gce instance setting

danieljbruce committed:
Eliminate the gcp cloud provider setting

danieljbruce committed:
run all tests

danieljbruce committed:
Eliminate duplicate project id from monitored

resource

danieljbruce committed:
Change the two metric names

danieljbruce committed:
Extend the timeout so that the exporter has

chance to work

danieljbruce committed:
Use spread syntax

danieljbruce committed:
Changed metric to otel instruments

danieljbruce committed:
Remove optional on Otel instruments

danieljbruce committed:
Rename initialize to getMetrics

danieljbruce committed:
Pin promisify to version 4

danieljbruce committed:
cast to string - compiler errors

danieljbruce committed:
Cast to string when building buffer

danieljbruce committed:
Remove a TODO - it is done

danieljbruce committed:
Delete the test that writes to metrics collected

danieljbruce committed:
eliminate unnecessary spacing

danieljbruce committed:
Remove console logs from the code

danieljbruce committed:
Eliminate the TODO in index

danieljbruce committed:
Add comment to mention mocha

danieljbruce committed:
Add another comment for code 0

danieljbruce committed:
Add a comment clarifying what the test does

danieljbruce committed:
Pin promisify to solve the linting issues

danieljbruce committed:
Rename test to ClientSideMetricsToMetricsHandler

Remove only

danieljbruce committed:
Rename another test suite

Client side metrics to GCM

danieljbruce committed:
Remove these console logs

danieljbruce committed:
Add guards to the after hooks on this test

danieljbruce committed:
Remove only

danieljbruce committed:
Increase the GCPMetricsHandler time

danieljbruce committed:
Fix samples test issue

danieljbruce committed:
Add a comment to test that works with the mock ser

danieljbruce committed:
Revert "Fix samples test issue"

This reverts commit 99c62718bb18533ac85e35f8dc72c727505d076f.

danieljbruce committed:
Increase the timeout

danieljbruce committed:
Catch failures to load ResponseParams

danieljbruce committed:
Change the latency buckets

Adjust the test fixtures as well to account for the latency bucket change.

danieljbruce committed:
Updated the resultCallback comment

danieljbruce committed:
Change the test description

danieljbruce committed:
Add a comment explaining what the test is about

danieljbruce committed:
Completed the test for two metrics handlers

danieljbruce committed:
remove only

danieljbruce committed:
Cast as histogram to make compile error go away

danieljbruce committed:
Move the duplicate copies of replaceTimestamps

into one file

danieljbruce committed:
Take steps to eliminate the ExportInput interface

danieljbruce committed:
Add a header

danieljbruce committed:
Remove only

danieljbruce committed:
Eliminate ExportInput and dependencies

danieljbruce committed:
Eliminate constant

danieljbruce committed:
Push conversion to status back to metrics collect

Also add if statement for timeseries.

danieljbruce committed:
Change property names to match MetricServiceClient

expectation

danieljbruce committed:
Solve the app_profile_id issue

danieljbruce committed:
It is actually app_profile not app_profile_id

danieljbruce committed:
Add guards that stop the export call

danieljbruce committed:
Remove only

danieljbruce committed:
Add a new test for ten metrics handlers

danieljbruce committed:
Remove only

danieljbruce committed:
Do not pass data through the Resource object

anymore

danieljbruce committed:
Add a test for writing duplicate points to MH

danieljbruce committed:
Eliminate interface

danieljbruce committed:
Set connectivity error count to 1 (not increment)

danieljbruce committed:
Merge branch '359913994-exporter-PR' of https://github.com/googleapis/nodejs-bigtable into 359913994-third-PR-CSM

# Conflicts:
#	src/client-side-metrics/operation-metrics-collector.ts

danieljbruce committed:
Edit the fixtures

danieljbruce committed:
Remove the guard

danieljbruce committed:
Change attempt status to string

danieljbruce committed:
Add a comment - undefined in conformance tests

danieljbruce committed:
Make metrics handlers per client

danieljbruce committed:
Add coment about latency

danieljbruce committed:
Wrap in try

danieljbruce committed:
Adjust the proxyquire mocks to mock handler in

right layer

danieljbruce committed:
Eliminate TODO

danieljbruce committed:
Eliminate TODO

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into 359913994-third-PR-CSM

# Conflicts:
#	src/client-side-metrics/exporter.ts
#	src/client-side-metrics/gcp-metrics-handler.ts
#	src/client-side-metrics/operation-metrics-collector.ts
#	test-common/expected-otel-export-input.ts
#	test/metrics-collector/typical-method-call.txt

danieljbruce committed:
Add some code so that first response latencies gets recorded

danieljbruce committed:
Run the linter

danieljbruce committed:
Reduce latency when collect metrics is false

danieljbruce committed:
Add a comment for the collect metrics variable

danieljbruce committed:
Remove unnecessary deltas

danieljbruce committed:
Add space back

danieljbruce committed:
Remove numRequests made increment

danieljbruce committed:
Check for existence of metrics

danieljbruce committed:
Finalize updates to do the test with 2 tables

danieljbruce committed:
Refactor the test setup into one file

danieljbruce committed:
Use the new setup table function

danieljbruce committed:
Remove only handlers

danieljbruce committed:
Add header

danieljbruce committed:
Add guard in the metrics collector

danieljbruce committed:
Revert "Remove the guard"

This reverts commit c581129e69eb6e9c2747554d69419837b878868f.

# Conflicts:
#	src/client-side-metrics/operation-metrics-collector.ts

danieljbruce committed:
Add the optimizations back again

danieljbruce committed:
cache cluster and zone so a call with 2 tables

works

danieljbruce committed:
Properly set up the table

danieljbruce committed:
Revert "cache cluster and zone so a call with 2 tables"

This reverts commit 873507aa21c2cfd376c0fd5093ca2e30bf8a3e0e.

danieljbruce committed:
Update the test setup

danieljbruce committed:
Test for multiple instances

danieljbruce committed:
Adapt the test to work with multiple clients

danieljbruce committed:
Remove only

danieljbruce committed:
Modification to handle no status

danieljbruce committed:
Increment number of exports

danieljbruce committed:
Add an error for the timeouts

danieljbruce committed:
Add a test for multiple simultaneous exports

danieljbruce committed:
Just test for one bigtable client right now

danieljbruce committed:
Name test differently so that we can distinguish

danieljbruce committed:
Clarify test responsibilities

Only fail in the timeout test if there are no exporting errors.

danieljbruce committed:
Eliminate variable assignment

danieljbruce committed:
Add a waiting script to the instance creation

danieljbruce committed:
Add the header and combine the tests

danieljbruce committed:
Remove only

danieljbruce committed:
Delete files that have been consolidated

danieljbruce commented:
We still need to do the QPS check so I'll come back here and comment when that is done.

danieljbruce committed:
Run the linter and add guards

danieljbruce committed:
Remove response params

danieljbruce committed:
Revert "Remove response params"

This reverts commit 6d9551cd1d8c4bc3ecb4c7530e51001854643530.

daniel-sanche commented on src/index.ts:
```diff
@@ -417,9 +424,19 @@ export class Bigtable {
   appProfileId?: string;
   projectName: string;
   shouldReplaceProjectIdToken: boolean;
+  clientUid = crypto.randomUUID();
```
I'm not seeing where this is used, but IIRC the client_uid is supposed to follow a specific structure: https://github.com/googleapis/python-bigtable/blob/987e3df4ce64a6079e03295f653363f1cc391e0b/google/cloud/bigtable/data/_metrics/handlers/opentelemetry.py#L144

andre-sampaio replied:
Reference in java as well: https://github.com/googleapis/java-bigtable/blob/main/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableExporterUtils.java#L113-L134

danieljbruce replied:
It gets passed in [here](https://github.com/googleapis/nodejs-bigtable/blob/c37a451c52ff5639fc9f5e76c88e93ba549fe6c9/src/client-side-metrics/operation-metrics-collector.ts#L116).

daniel-sanche commented on src/index.ts:
```diff
@@ -517,6 +534,16 @@ export class Bigtable {
+    } else {
+      this.collectMetrics = true;
+      this.metricsHandlers = [
+        new GCPMetricsHandler(new CloudMonitoringExporter()),
```
is there a reason to have `CloudMonitoringExporter` as an argument? Will a different one ever be passed in?

danieljbruce replied:
Within the client library code `CloudMonitoringExporter` is always used, but tests sometime pass in a mock exporter so that we can unit test GCPMetricsHandler's behaviour. If you want we can define a DefaultGCPMetricsHandler class that just always uses the CloudMonitoringExporter to eliminate the confusion.

daniel-sanche replied:
It looks like [we discussed this in the other PR](https://github.com/googleapis/nodejs-bigtable/pull/1583#discussion_r2034178097). But this one probably makes more sense

Isn't there a way to inject the mock without making the internal variable part of the API surface?

danieljbruce replied:
In multiple tests a mock exporter is passed in here:

![image](https://github.com/user-attachments/assets/6d928ae0-48c1-4ee1-a947-706dc0785d1f)

We could inject the mock with proxyquire, doing something like https://github.com/googleapis/nodejs-bigtable/pull/1571/commits/1ddaa7df1d142abff4fe399a2da6fa8b1cf00aaf for each test, but when proxyquire is involved it can be hard to debug later. What do you want to do here?

daniel-sanche commented on src/index.ts:
```diff
@@ -417,9 +424,19 @@ export class Bigtable {
+  // user provides a `false` value and opts out of metrics collection that
+  // the metrics collector is ignored altogether to reduce latency in the
+  // client.
+  collectMetrics: boolean;
```
nit: Using a verb for a flag can be a bit confusing, it sounds more like a function name. I'd suggest something like `metricsEnabled`

daniel-sanche commented on src/tabular-api-surface.ts:
```diff
@@ -507,6 +524,24 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
         return false;
       };
 
+      if (this.bigtable.collectMetrics) {
```
Could we have it do the callbacks every time, regardless of this variable?

We should assume almost all users have CSM enabled, since that's the default. And we'd be introducing a lot of extra branching if we check the metrics collection state every time

danieljbruce replied:
Sure. I added the check out of concern for latency, but it is likely pretty minor.

daniel-sanche commented on src/tabular-api-surface.ts:
```diff
@@ -532,6 +567,10 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
               numConsecutiveErrors,
               backOffSettings
             );
+            metricsCollector?.onAttemptComplete(
```
I see a `return` on line 553. Don't you need to finalize things there?

danieljbruce replied:
Good catch. Added that fix!

daniel-sanche commented on src/tabular-api-surface.ts:
```diff
@@ -547,16 +586,33 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
               //
               error.code = grpc.status.CANCELLED;
             }
+            metricsCollector?.onAttemptComplete(
```
I REALLY don't like the comment above this: 

> According to Gemini, the appropriate code for a closed client is CANCELLED since the user actually cancelled the call by closing the client.

You're responsible for the code, not Gemini. What do you think is the appropriate code for a closed client?

daniel-sanche replied:
Also, I don't like the idea of referencing a specific test in the code here. We don't just want the test to pass, we want to make sure the underlying requirements are being met

daniel-sanche commented on src/tabular-api-surface.ts:
```diff
@@ -547,16 +586,33 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
+          metricsCollector?.onAttemptComplete(
+            this.bigtable.projectId,
+            grpc.status.OK
+          );
```
FWIW, in Python I created a separate `onOperationSucceeded` helper to call onAttemptComplete and onOpeationComplete with grpc.status.OK. This makes it a bit easier to see the happy-path, and can simplify things a bit. Up to you if thats helpful here

danieljbruce replied:
I added a onLastAttemptCompleted method which simplifies things in a few places. It will reduce the amount of code necessary for methods that use the metrics collector.

andre-sampaio commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -17,10 +17,23 @@ import {IMetricsHandler} from './metrics-handler';
-const ResponseParams = root.lookupType('ResponseParams');
+
+let ResponseParams: gax.protobuf.Type | null;
+try {
```
I'm a bit concerned with failing silently here due to a test failing. If the tests that are failing are only the tests for the samples I'd rather skip those tests in the CI until we figure out what to do

danieljbruce replied:
The trouble is that the error is for the entire suite so we would have to skip the entire CI check in order for this to be unblocked.

andre-sampaio replied:
Is it only for the [samples](https://github.com/googleapis/nodejs-bigtable/tree/main/samples/test) test suite though? If it is I would still rather just skip those to be honest

andre-sampaio replied:
If we really don't want to skip can we gate this behind an env variable that is set from the test?

danieljbruce replied:
Yes. Only for the samples suite.

We can change the npm script to do `cd samples/ && npm link ../ && npm install && npx cross-env METRICS_DEBUG=true npm test && cd ../` instead of `cd samples/ && npm link ../ && npm install && npm test && cd ../` and that might work. Do you want to do this?

andre-sampaio commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -139,75 +156,88 @@ export class OperationMetricsCollector {
+          MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS;
+        this.attemptCount++;
+        const endTime = new Date();
+        if (projectId && this.attemptStartTime) {
```
We expect this to always be true right? If so can we log an error if this happens?

danieljbruce replied:
Done. I added an else with a warning

andre-sampaio commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -139,75 +156,88 @@ export class OperationMetricsCollector {
+          this.state =
+            MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_SOME_ROWS_RECEIVED;
+          const endTime = new Date();
+          if (projectId && this.operationStartTime) {
```
We expect this to always be true right? If so can we log an error if this happens? Same note for other similar checks where we expect the condition to be always be true

danieljbruce replied:
Now logging an error if this happens

andre-sampaio commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -218,33 +248,40 @@ export class OperationMetricsCollector {
-            }
-          });
+    try {
+      if (
```
Can we log an error if this is false?

danieljbruce replied:
Yup. Done

andre-sampaio commented on src/tabular-api-surface.ts:
```diff
@@ -547,16 +586,33 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
               //
               error.code = grpc.status.CANCELLED;
             }
+            metricsCollector?.onAttemptComplete(
```
Instead of calling on attempt complete in both the `if` and `else` conditions we can just move it out of the block

danieljbruce replied:
onAttemptComplete is no longer called in both places after the introduction of the `onLastAttemptCompleted` method.

danieljbruce committed:
Remove the samples test workaround

danieljbruce committed:
Add a wrapper for the try/catch logic

danieljbruce commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -17,10 +17,23 @@ import {IMetricsHandler} from './metrics-handler';
+   * directory, open 'protos/google/bigtable/v2/response_params.proto'. Since
+   * these tests don't use this module we can suppress the error for now to
+   * unblock the CI pipeline.
+   */
```
[here is a log](https://btx.cloud.google.com/invocations/d22777c4-0b03-4939-a47d-a9a3762c7691/targets/cloud-devrel%2Fclient-libraries%2Fnodejs%2Fpresubmit%2Fgoogleapis%2Fnodejs-bigtable%2Fnode14%2Fsamples-test;config=default/log) that shows what happens without this change. Not a problem with a specific test, just an error in the 'fs' module all of a sudden. It might go away when we start running these tests with a higher version of Node.

daniel-sanche replied:
is this still relevant? It looks like that CL was rolled back

danieljbruce replied:
[the most recent log](https://btx.cloud.google.com/invocations/c3d9117f-67da-4c7c-b522-bb5707a1ce38/targets/cloud-devrel%2Fclient-libraries%2Fnodejs%2Fpresubmit%2Fgoogleapis%2Fnodejs-bigtable%2Fnode14%2Fsamples-test/log) shows this is still an issue so it must be due to a fs upgrade or something. I don't know why this is suddenly cropping up.

danieljbruce committed:
Revert "Remove the samples test workaround"

This reverts commit 0164f557f4e8ae37f2b14868abfcfcaac880c935.

# Conflicts:
#	src/client-side-metrics/operation-metrics-collector.ts

danieljbruce committed:
Spelling correction

danieljbruce committed:
Add a helper method for checking the state

danieljbruce committed:
this.state

danieljbruce commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -123,13 +136,17 @@ export class OperationMetricsCollector {
-    } else {
-      console.warn('Invalid state transition');
+    try {
+      if (this.state === MetricsCollectorState.OPERATION_NOT_STARTED) {
```
Okay. This is now handled with a wrapper `checkState` function.

danieljbruce committed:
Optional statuses

danieljbruce committed:
Eliminate old comments

danieljbruce commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -139,75 +156,88 @@ export class OperationMetricsCollector {
+          this.metricsHandlers.forEach(metricsHandler => {
+            if (metricsHandler.onAttemptComplete) {
+              // attemptStatus?.toString() is optional because in a test proxy
+              // test the server does not send back the status.
```
Ok. I eliminated the optional and the tests seem to work so I'm guess this observation was just a mistake. I changed it back now.

danieljbruce commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -218,33 +248,40 @@ export class OperationMetricsCollector {
+            this.metricsHandlers.forEach(metricsHandler => {
+              if (metricsHandler.onOperationComplete) {
+                // finalOperationStatus?.toString() is optional because in a test
+                // proxy test the server does not send back the status.
```
Ok. I eliminated the optional and the tests seem to work so I'm guess this observation was just a mistake. I changed it back now.

danieljbruce commented on src/index.ts:
```diff
@@ -417,9 +424,19 @@ export class Bigtable {
   static AppProfile: AppProfile;
   static Instance: Instance;
   static Cluster: Cluster;
+  // Each time a metrics handler is created it introduces significant latency.
```
In our other PR we are considering changing the code to initialize the stack 1 time so let's resolve that discussion and then decide what to do here next.

danieljbruce replied:
I did some measurements [here](https://paste.googleplex.com/6658498605350912) on `main-readrows-latency-experiment`. It honestly seems like the latency is negligible compared to the client initialization time, its just that it would add up a lot if we initialized a stack for every table or every call.

daniel-sanche replied:
Maybe worth changing this comment then if latency isn't the reason

danieljbruce replied:
Latency would be a reason if we created a stack for every readrows call. Too many calls to the exporter is another reason. I've updated the comment.

danieljbruce replied:
Actually, we don't need this on the client anymore because the OTEL stacks now belong to a singleton. So I moved the metrics handlers to live as a singleton on the OperationMetricsCollector instead of on the client here.

We shouldn't be exposing a `metricsHandlers` variable on the client if we don't have to. This is a public interface.

danieljbruce committed:
Add a generateClientUuid function

danieljbruce committed:
Changed the name to metrics enabled

danieljbruce committed:
End call on ignored status code

danieljbruce committed:
eliminate mention of gemini

danieljbruce committed:
Add onOperationSucceeded helper

danieljbruce committed:
Change name to last attempt completed

danieljbruce committed:
Use onLastAttemptCompleted here

danieljbruce committed:
run linter

danieljbruce committed:
Add handlers no matter what

danieljbruce committed:
Add warning if check fails

danieljbruce committed:
Add console warns for various checks

danieljbruce committed:
Use the hrtime module instead of the Date module

# Conflicts:
#	src/client-side-metrics/operation-metrics-collector.ts
#	test/metrics-collector/metrics-collector.ts

daniel-sanche commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -58,6 +75,33 @@ enum MetricsCollectorState {
+  if (allowedStates.includes(currentState)) {
+    return fn();
+  } else {
+    console.warn('Invalid state transition');
```
Did you intend to have this always log, or should this also be behind METRICS_DEBUG?

danieljbruce replied:
I ended up gating this with an environment variable and actually throwing an error when the variable is true. We have to change the npm script in package json to use the variable in tests:

ie. npx cross-env METRICS_DEBUG=true mocha build/system-test --timeout 600000

Should we use the variable in tests too?

daniel-sanche commented on src/tabular-api-surface.ts:
```diff
@@ -507,6 +524,22 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
+          }) => {
+            metricsCollector?.onStatusMetadataReceived(status);
+          }
+        );
```
I imagine you'll need to do something like this for every rpc? Should there be a helper for this?

danieljbruce replied:
Yes. I added `handleStatusAndMetadata` which should help us here.

danieljbruce committed:
Remove the try block

danieljbruce committed:
Modified comment

danieljbruce committed:
Modified export comment

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into 359913994-third-PR-CSM

andre-sampaio commented on src/tabular-api-surface.ts:
```diff
@@ -605,11 +602,7 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
         })
         .on('end', () => {
```
Won't this be called on errors as well?

danieljbruce replied:
No. I did an experiment on https://github.com/googleapis/nodejs-bigtable/commit/e9349762dced971590fe07f1752c6fd70ebbf085 and I get the output:

```
  Bigtable/ReadRows
error row stream happened
error reached user
    ✔ should create read stream and read synchronously (220ms)
```

danieljbruce committed:
Make the metrics handler static on the MC

MC = OperationMetricsCollector

danieljbruce committed:
Throw an error for better testing

Also. Always gate the error throwing to just tests.

danieljbruce committed:
Add a helper for adding handlers

danieljbruce committed:
Directionally correct proxyquire stuff

danieljbruce committed:
Remove traces and various comments

danieljbruce committed:
Set instance name

danieljbruce committed:
Factor out the fake bigtable creator

danieljbruce committed:
Remove only

danieljbruce committed:
Eliminate onLastAttemptCompleted

danieljbruce commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -218,34 +280,53 @@ export class OperationMetricsCollector {
+   */
+  onLastAttemptCompleted(projectId: string, attemptStatus: grpc.status) {
+    this.onAttemptComplete(projectId, attemptStatus);
+    this.onOperationComplete(projectId, attemptStatus);
```
Okay. I eliminated onLastAttemptCompleted.

danieljbruce committed:
eliminate environment variable setting

daniel-sanche commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -58,6 +72,35 @@ enum MetricsCollectorState {
+  } else {
+    if (METRICS_DEBUG) {
+      throw Error('Invalid state transition');
+    }
```
Since you're throwing errors now, this can be simplified quite a bit:
1. remove the `if (METRICS_DEBUG)` check here, and always throw an error if it's in the wrong state. `withMetricsDebug` will catch it and decide whether to raise. That keeps all the `METRICS_DEBUG` logic in one place
2. You now have the option to remove the whole `fn: () => T` callback part, and just return if it's in a valid state. Execution will just continue if the state is good, or end here if it's in a bad state. (But I know JS is more callback-centric than other languages, so maybe it's more idiomatic to keep it as-is. I find lots of nested callbacks hard to reason about, but maybe that's just me)

danieljbruce replied:
> I find lots of nested callbacks hard to reason about, but maybe that's just me

I do too. I think this is a good suggestion.

> withMetricsDebug will catch it and decide whether to raise

This is true, but it does mean we should only use checkState when it is wrapped in withMetricsDebug.

daniel-sanche replied:
> This is true, but it does mean we should only use checkState when it is wrapped in withMetricsDebug.

Wasn't that the intention? That you wrap everything that could call in `withMetricsDebug`, and can then surpress exceptions?

daniel-sanche commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -13,10 +13,24 @@
+ * directory, open 'protos/google/bigtable/v2/response_params.proto'. Since
+ * these tests don't use this module we can suppress the error for now to
+ * unblock the CI pipeline.
+ */
```
Is this comment still relevant?

danieljbruce replied:
Unfortunately, this still comes up even after the Node 18 upgrade: https://btx.cloud.google.com/invocations/b88b3b6c-0ace-413b-9b8c-c9bef9eceb03/targets/cloud-devrel%2Fclient-libraries%2Fnodejs%2Fpresubmit%2Fgoogleapis%2Fnodejs-bigtable%2Fnode18%2Fsamples-test/log

danieljbruce replied:
We'll see if using the path module for relative path resolution fixes the problem....

daniel-sanche commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -139,76 +209,92 @@ export class OperationMetricsCollector {
+              }
+            );
+          } else {
+            console.warn('ProjectId and start time should always be provided');
```
should this be a throw?

It seems to me like all metrics errors should be handled with the METRICS_DEBUG logic. But maybe you wanted different levels for exceptions vs warnings?

danieljbruce replied:
Yes. I think throwing here keeps the logic simpler so I made this change.

daniel-sanche commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -139,76 +209,92 @@ export class OperationMetricsCollector {
+        () => {
+          this.state =
+            MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET;
+          this.attemptStartTime = new Date();
```
I thought I saw a change to `hrtime.bigint` for timestamps. Should that be done here too?

daniel-sanche commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -139,76 +209,92 @@ export class OperationMetricsCollector {
+          () => {
+            this.state =
+              MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_SOME_ROWS_RECEIVED;
+            const endTime = new Date();
```
same comment about `hrtime.bigint` timestamps

danieljbruce replied:
This change has been cherry picked.

danieljbruce committed:
thread projectId through test framework

danieljbruce committed:
Add error handling

danieljbruce committed:
Test with different project

danieljbruce committed:
Localize the MetricServiceClient

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

danieljbruce committed:
remove callback on checkState

danieljbruce committed:
Throw instead of console.warn

danieljbruce committed:
Merge branch '359913994-third-PR-CSM' of https://github.com/googleapis/nodejs-bigtable into 359913994-third-PR-CSM

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into 359913994-third-PR-CSM

# Conflicts:
#	protos/protos.d.ts
#	src/client-side-metrics/operation-metrics-collector.ts
#	test/metrics-collector/metrics-collector.ts

danieljbruce committed:
run the linter

danieljbruce committed:
Fix failing tests

Have to change the static variable on the mock instead

danieljbruce committed:
Add promise catching

Fixes the linter errors

danieljbruce committed:
Remove unused imports

danieljbruce committed:
Remove unused import

danieljbruce committed:
Use the path module

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into 359913994-third-PR-CSM

# Conflicts:
#	src/client-side-metrics/operation-metrics-collector.ts
#	test/metrics-collector/metrics-collector.ts

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

daniel-sanche commented on src/client-side-metrics/exporter.ts:
```diff
@@ -328,7 +342,8 @@ export class CloudMonitoringExporter extends MetricExporter {
           },
         );
-        await this.monitoringClient.createTimeSeries(
+        const monitoringClient = new MetricServiceClient({projectId});
```
we may need to authenticate for this project Id

Since we have a bug tracking this, we can probably address this later

danieljbruce committed:
Delete application latencies from the check

danieljbruce committed:
Get rid of bigtable singleton

The bigtable singleton makes it really confusing to read the code. Instead, let’s have a unique bigtable client for each test and mock it out each time as necessary so that we know the exact state of the client in each test.

danieljbruce committed:
Merge branch 'main' into 359913994-third-PR-CSM

daniel-sanche reviewed (changes requested):
It looks like we have to make some architecture changes to properly pass the auth token down to the MetricsServiceClient

mutianf commented on src/tabular-api-surface.ts:
```diff
@@ -540,23 +567,31 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
               //
               error.code = grpc.status.CANCELLED;
             }
+            metricsCollector?.onOperationComplete(
```
where are the other metric labels recorded? e.g. instance id, table id, method, etc

danieljbruce committed:
fix: Get rid of the global singletons thereby improving client side metrics design (#1605)

* added metrics controller at the client level

* renamed class

* did some clean up around handler creation

* fixed accidental deletion

* use options instead of auth

* fixed typo

* don't pass project on each call

* call getProjectId when setting up config manager

* pulled out project id from metrics model

* Fix: Resolve CSM branch compilation and auth issues

This commit addresses compilation errors in the `csm_sanche` branch
and ensures that authentication credentials can be passed to the
`CloudMonitoringExporter`.

The `csm_sanche` branch aimed to remove global singletons for client-side
metrics. This change continues that effort by:

- Fixing type errors and import issues in `ClientSideMetricsConfigManager`,
  `GCPMetricsHandler`, and `CloudMonitoringExporter`.
- Modifying `CloudMonitoringExporter` and `GCPMetricsHandler` to accept
  an `options` object in their constructors. This object is passed down
  to the `MetricServiceClient`, allowing credentials to be provided.
- Updating test files to align with these changes.
- Adding new unit tests for `ClientSideMetricsConfigManager.getGcpHandlerForProject`
  to verify:
    - Correct instantiation of `GCPMetricsHandler`.
    - Propagation of the `options` object (for credentials).
    - Caching behavior of `GCPMetricsHandler` instances per project ID.

These changes ensure that the client-side metrics functionality
compiles correctly and that the exporter can be properly authenticated
for sending metrics to Google Cloud Monitoring.

* run the linter

* Various changes to clarify types

* Parameter should be in this position

* Pass in project Id first

* Name of the project

* Go back to fetching OTEL instruments for each call

* expect project id in the metrics collector

* Make small change moving config manager

Fix the tests because the mock has moved too

* Debug current test of interest

* Thread the projectId through the OTEL instruments

again

* Run the linter

* Use metrics enabled to conditionally use metrics c

Turn on metrics enabled
Eliminate unnecessary parameter pass thrus

* Don’t store options

* Remove set options

* Separate object for creating metrics collector

Also adjust mocking for the system test.

* remove the console logs

* Only create one exporter per test

* Remove console trace

* Add fn to get the handler from the exporter

* Add a test making sure the exporter gets the opts

* Provide the proper mock for the metric service cli

client

* Each handler has a different instrument stack now

So these tests need to change

* cosmetic changes

Add the check for the projectId again?

* Fix compiler error for projectId

* Address most of the comments Jules added

* Remove unnecessary comments

* Address header check

* Remove unused imports

* Removed automatically added comments

* Pass in the projectId at constructor time

* Fix the tests so they pass

* Fix the manager tests

* Move the instrument stack cache

* Eliminate some proxyquire mocks

* More changes required to bring

handlers to the client

* Resolve the remaining compile time errors

* Fix the tests for createReadStream

* Fix Should export a value ready for sending to the

* Fix test due to metrics handler contract

* Remove only

* Add header

* Change fake bigtable mock

* Eliminate the assertion checks inside handler

* Fix the projectId test

* Remove onlys

* Fix the hundred metrics handlers test

* Add a test for 100 clients

* Add retries to metric service client

* Add a timeout to the hundred handlers

* Remove console logs

* Better test for many clients

* No set console log

* Eliminate hundred metrics handlers test

* Fix issues with system tests due to new design

* Remove onlys

* separate variable for handlers

---------

Co-authored-by: Daniel Sanche <d.sanche14@gmail.com>
Co-authored-by: google-labs-jules[bot] <161369871+google-labs-jules[bot]@users.noreply.github.com>

danieljbruce commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -59,6 +69,31 @@ enum MetricsCollectorState {
+    return fn();
+  } catch (e) {
+    if (METRICS_DEBUG) {
+      throw e;
```
Yes it will. But only when the METRICS_DEBUG variable is enabled.

danieljbruce replied:
This logs a warning now anyways

danieljbruce committed:
Add warnings

danieljbruce committed:
private readonly variable

danieljbruce committed:
Add default zone and cluster ids

danieljbruce commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -59,6 +69,31 @@ enum MetricsCollectorState {
+  } catch (e) {
+    if (METRICS_DEBUG) {
+      throw e;
+    }
```
Sure. Logged a warning here instead.

danieljbruce commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -334,29 +393,33 @@ export class OperationMetricsCollector {
-        ).clusterId;
+    withMetricsDebug(() => {
+      if (!this.zone || !this.cluster) {
+        const INSTANCE_INFORMATION_KEY = 'x-goog-ext-425905942-bin';
```
Sure. Moved to global variable

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into 359913994-third-PR-CSM

# Conflicts:
#	system-test/service-path.ts

danieljbruce committed:
Add a test for second project id

danieljbruce committed:
Better naming conventions for project

danieljbruce commented on src/client-side-metrics/exporter.ts:
```diff
@@ -326,9 +323,11 @@ export class CloudMonitoringExporter extends MetricExporter {
         );
-        await this.monitoringClient.createTimeSeries(
+
+        await this.client.createTimeSeries(
```
I don't think so. Keep in mind that this is just the MetricServiceClient used for writing the metrics. What is your concern?

daniel-sanche replied:
Yeah, it [looks like it does exist](https://github.com/googleapis/google-cloud-node/blob/b41f79df5e4f9472761ce1383d4d85785c647a15/packages/google-cloud-monitoring/src/v3/metric_service_client.ts#L1207)

IIRC, createTimeSeries is more generic, but createServiceTimeSeries is meant for GCP metrics

danieljbruce commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -13,13 +13,22 @@
+
+// When this environment variable is set then print any errors associated
+// with failures in the metrics collector.
+const METRICS_DEBUG = process.env.METRICS_DEBUG;
```
No. Default to false. We don't want to flood the customer's terminal with warnings if there is some kind of mistake?

danieljbruce replied:
> We should have integration test to make sure the instrumentation to publish metrics are correct

There are tons of tests that do this like https://github.com/googleapis/nodejs-bigtable/blob/f8fd4be1d5f415b90739090bbd4ee6d1a2cfa021/system-test/gcp-metrics-handler.ts#L238 for example. It would have been impossible to develop the instrumentation without them. See:
https://github.com/googleapis/nodejs-bigtable/pull/1583
https://github.com/googleapis/nodejs-bigtable/pull/1569
https://github.com/googleapis/nodejs-bigtable/pull/1566

So you want me to set this variable to true?

danieljbruce committed:
don’t retry MSC, warning

danieljbruce commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -59,10 +70,37 @@ enum MetricsCollectorState {
   OPERATION_COMPLETE,
 }
 
+// This method swallows errors when metrics debugging is not enabled so
```
Right. Changed

danieljbruce commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -59,10 +70,37 @@ enum MetricsCollectorState {
+  } catch (e) {
+    if (METRICS_DEBUG) {
+      console.warn('METRICS_DEBUG warning');
+      console.warn((e as ServiceError).message);
```
I don't think it will include the stack trace. Just something like https://github.com/googleapis/nodejs-bigtable/blob/main/system-test/service-path.ts#L191.

danieljbruce committed:
Change to unspecified

danieljbruce committed:
Remove unused method

danieljbruce commented on src/index.ts:
```diff
@@ -128,6 +140,13 @@ function getDomain(prefix: string, opts?: gax.ClientOptions) {
   }`;
 }
 
+function generateClientUuid() {
```
Yes. This is not used at all.

danieljbruce commented on src/tabular-api-surface.ts:
```diff
@@ -332,14 +336,24 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
+    const metricsCollector =
+      this.bigtable._metricsConfigManager.createOperation(
+        MethodName.READ_ROWS,
+        StreamingState.STREAMING,
```
You're right. `row.get` calls `createReadStream` here, but remember that createReadStream always hits the readRows endpoint. In fact, I don't even see a gapic method for readRow.

If a user calls `row.get` then do we want to log the metrics as `READ_ROW` because the user is just getting a single row or do we want to log it as `READ_ROWS` because really it is hitting the READ_ROWS grpc endpoint?

mutianf replied:
If user calls row.get, we want to log the method as ReadRow and Streaming = false.

danieljbruce replied:
https://github.com/googleapis/nodejs-bigtable/pull/1612 is the change required to complete this work.

daniel-sanche replied:
@mutianf "ReadRow and Streaming = False"? I thought it was "ReadRow**s** and Streaming = False". And looking at go's implementation,[ I don't see ReadRow being used anywhere](https://github.com/search?q=repo%3Agoogleapis/google-cloud-go%20gaxInvokeWithRecorder&type=code). But let me know if I'm misunderstanding

daniel-sanche replied:
Oh wait, [I dug up this conversation](https://github.com/googleapis/python-bigtable/pull/923#discussion_r1473408291) in the Python PR. Igor said:

>I think java ReadRow tag was my mistake :(
> Lets keep this inline with server metrics. I think the correct path forward will be stanley's request profiles

So let me know whether what you think

danieljbruce replied:
This is more of a requirements question so I wouldn't know.

Right now, `table.getRows` and `table.createReadStream` both record for `READ_ROWS` and `STREAMING`
`row.get` records for `READ_ROW` and `Unary`.

@mutianf Let me know if that's correct.

danieljbruce commented on src/tabular-api-surface.ts:
```diff
@@ -540,23 +557,25 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
               //
               error.code = grpc.status.CANCELLED;
             }
+            metricsCollector.onOperationComplete(error.code);
```
Attempt complete is already part of operation complete https://github.com/googleapis/nodejs-bigtable/pull/1571/files#diff-10f88e720bed4e5a35d5d0e992011fd658dc800b0ff6510fd4ef6137f036ae29R288.

danieljbruce commented on src/tabular-api-surface.ts:
```diff
@@ -540,23 +557,25 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
         })
         .on('end', () => {
           activeRequestStream = null;
+          metricsCollector.onOperationComplete(grpc.status.OK);
```
Attempt complete is already part of operation complete https://github.com/googleapis/nodejs-bigtable/pull/1571/files#diff-10f88e720bed4e5a35d5d0e992011fd658dc800b0ff6510fd4ef6137f036ae29R288.

danieljbruce committed:
Merge branch 'main' into 359913994-third-PR-CSM

mutianf commented on system-test/metric-service-client-credentials.ts:
```diff
@@ -0,0 +1,91 @@
+  });
+  it('should use second project for the metric service client', async () => {
+    const SECOND_PROJECT_ID = 'second-project-id';
+    const clientOptions = {metricsEnabled: true, projectId: SECOND_PROJECT_ID};
```
is the SECOND_PROJECT_ID here the project id of the bigtable client?

danieljbruce replied:
Yes. I added a comment explaining what this test intends to do. The idea is that if the second project is provided to the Bigtable client then when the exporter uses the options it gets to build the metric service client the metric service client should use the right project.

mutianf reviewed (commented):
Do we have a test that tests the actual value we're recording to make sure the measurement is correct?

danieljbruce committed:
feat: Record READ_ROW metrics for single row calls (#1612)

* Delegate createReadStream

* Pull out getRows into an internal function

* Remove only

* We have some duplicate code so add a TODO

* Almost fixed all the row tests

* Fix last remaining test

* Settings for Read_Row vs Read_Rows

* Set to false singlerow

* Add new READ_ROW test

Also pull check into a separate fn

* Fix the single row call for the test

* Fix test for single row read

* Refactor some of the proxyquire code

* Remove only

danieljbruce committed:
Added a comment for the test

danieljbruce committed:
Include credentials in the test

danieljbruce commented on system-test/client-side-metrics.ts:
```diff
@@ -0,0 +1,514 @@
+  describe('Bigtable/ClientSideMetricsToGCMTimeout', () => {
+    // This test suite simulates a situation where the user creates multiple
+    // clients and ensures that the exporter doesn't produce any errors even
+    // when multiple clients are attempting an export.
```
> This could happen if multiple clients share the same UUID and publishes metric data points too frequently.

Yes. There has been lots of discussion about the maximum sampling error and when it occurs. We even thought we needed design changes due to this issue, but then reverted such changes because the client uuids are different for every client and this is enough to avoid the maximum sampling error. `should send the metrics to Google Cloud Monitoring for a ReadRows call with thirty clients` test captures this multiple client case.

danieljbruce commented on system-test/client-side-metrics.ts:
```diff
@@ -0,0 +1,514 @@
+            const table = instance.table(tableId1);
+            await table.getRows();
+            const table2 = instance.table(tableId2);
+            await table2.getRows();
```
I'm open to checking metrics publishing programmatically though. I'll need to research more on how to do that and it could be a time investment.

danieljbruce committed:
Modify test to pass credentials in the preferred

manner

danieljbruce committed:
Remove the calls that aren’t needed from unit test

danieljbruce committed:
Add a check for published metrics

danieljbruce committed:
Change to check for a custom endpoint

danieljbruce committed:
apiEndpoint should be part of the metrics handler

daniel-sanche commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -59,10 +70,36 @@ enum MetricsCollectorState {
  * A class for tracing and recording client-side metrics related to Bigtable operations.
  */
 export class OperationMetricsCollector {
+  private readonly INSTANCE_INFORMATION_KEY = 'x-goog-ext-425905942-bin';
```
a comment here might be helpful, since it looks pretty random without context

daniel-sanche commented on src/index.ts:
```diff
@@ -16,9 +16,15 @@ import {replaceProjectIdToken} from '@google-cloud/projectify';
+} from 'google-gax';
 import * as gax from 'google-gax';
 import * as protos from '../protos/protos';
+import * as os from 'os';
```
looks like this isn't needed?

danieljbruce replied:
Removed. At this point there may be some automation or a bot we can set up for little changes like this. It could be a real time saver.

daniel-sanche commented on src/utils/createReadStreamInternal.ts:
```diff
@@ -0,0 +1,445 @@
+ */
+export function createReadStreamInternal(
+  table: TabularApiSurface,
+  singleRow: boolean,
```
I don't know if changing this would be a lot of work, but I think it would be cleaner to pass a `metricsCollector` object here directly, instead of passing a bool and then using that to build different `metricsCollector` instances

danieljbruce replied:
Yes. This is better and I added this change.

daniel-sanche commented on src/utils/getRowsInternal.ts:
```diff
@@ -0,0 +1,66 @@
+ * @example <caption>include:samples/api-reference-doc-snippets/table.js</caption>
+ * region_tag:bigtable_api_get_rows
+ */
+export function getRowsInternal(
```
nit: It seems a but weird to have this helper that basically just defers to another helper, and is only used in a few places. I'd probably suggest to move this in with createReadStreamInternal (if not cut it out completely), since it seems like a pretty thin wrapper over it. I'm not sure if that would break a lot of tests or anything though

danieljbruce replied:
These changes were introduced in https://github.com/googleapis/nodejs-bigtable/pull/1612.

This is necessary and here is why:
`table.getRows` calls `getRowsInternal`.
`table.createReadStream` calls `createReadStreamInternal`.

`getRowsInternal` and `createReadStreamInternal` both need a parameter telling them if the call is for a single row or for a set of rows. We cannot add this parameter to `table.getRows` or `table.createReadStream` because it changes the API surface.

danieljbruce replied:
If `createReadStreamInternal` was pasted inline into `getRowsInternal` then `createReadStream` would have nothing to call.

daniel-sanche reviewed (changes requested):
I left a new more comments. Along with those, it looks like there are these conversations that seem unresolved:
- createServiceTimeSeries: [https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2114495885](https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2114495885)
- remove retries? [https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2114496978](https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2114496978)
- READ_ROW vs READ_ROWS [https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2114556372](https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2114556372)
- Exceptions [https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2124659124](https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2124659124)
- Default METRICS_DEBUG to true? [https://github.com/googleapis/nodejs-bigtable/pull/1571/files#r2114536705](https://github.com/googleapis/nodejs-bigtable/pull/1571/files#r2114536705)

Overall, it's looking good!

daniel-sanche commented on src/client-side-metrics/operation-metrics-collector.ts:
```diff
@@ -13,13 +13,22 @@
+const protoPath = path.join(
+  __dirname,
+  '../../protos/google/bigtable/v2/response_params.proto',
 );
```
She also said:
>The main thing is it might be good to see if you can pull the parsed protos via gax or gapic-loaded versions. The parsing is expensive and sometimes leads to memory leakage if it happens a lot. (The text of the protos will get stuck for some reason.)

Let me know if you think there's a way to load them directly from gapic, because that would be cleaner. But I'm not familiar enough with the node stack to say much more than that

danieljbruce committed:
Remove import

danieljbruce committed:
Add a comment about the key

danieljbruce committed:
Move generateUUID function

danieljbruce committed:
Eliminate generateClientUuid and leftovers

danieljbruce commented on src/client-side-metrics/generate-client-uuid.ts:
```diff
@@ -0,0 +1,39 @@
+  const currentPid = process.pid || '';
+  const uuid4 = crypto.randomUUID();
+  return `node-${uuid4}-${currentPid}${hostname}`;
+}
```
I implemented the alternative I suggested so that we could move this function.

danieljbruce committed:
Pass in metrics collector

danieljbruce committed:
Use more specific method

danieljbruce commented on system-test/client-side-metrics.ts:
```diff
@@ -0,0 +1,514 @@
+              await table2.getRows();
+            }
+          }
+        } catch (e) {
```
I mention the test function swallowing the error because async functions swallow errors.

This comment is not about exporter errors. The catch block would not catch exporter errors.

For exporter errors it looks like the error gets logged to the console.

For exporter errors we have this comment in the code:
// The resultCallback typically accepts a value equal to {code: x}
// for some value x along with other info. When the code is equal to 0
// then the operation completed successfully. When the code is not equal
// to 0 then the operation failed. Open telemetry logs errors to the
// console when the resultCallback passes in non-zero code values and
// logs nothing when the code is 0.

Actually I just did an experiment on https://github.com/googleapis/nodejs-bigtable/compare/359913994-third-PR-CSM-experiment?expand=1 and it looks like it doesn't log anything to the console at all. We have to correct this comment. Do we want errors to get printed to the console?

danieljbruce replied:
Corrected the comment

danieljbruce committed:
Correct the comment

danieljbruce commented:
> I left a new more comments. Along with those, it looks like there are these conversations that seem unresolved:
> 
> * createServiceTimeSeries: [feat: Add client side metrics collection to readRows calls #1571 (comment)](https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2114495885)
> * remove retries? [feat: Add client side metrics collection to readRows calls #1571 (comment)](https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2114496978)
> * READ_ROW vs READ_ROWS [feat: Add client side metrics collection to readRows calls #1571 (comment)](https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2114556372)
> * Exceptions [feat: Add client side metrics collection to readRows calls #1571 (comment)](https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2124659124)
> * Default METRICS_DEBUG to true? https://github.com/googleapis/nodejs-bigtable/pull/1571/files#r2114536705
> 
> Overall, it's looking good!

Do we want to set METRICS_DEBUG to true?

daniel-sanche reviewed (dismissed):
It looks like my concerns have all been addressed. The two remaining questions for Mattie that I see are:
- do we want to record the read_row metric as READ_ROWS (like Java), or READ_ROW (like Igor suggested for Python)? (https://github.com/googleapis/nodejs-bigtable/pull/1571#discussion_r2114556372)
- Should METRICS_DEBUG default on or off? (https://github.com/googleapis/nodejs-bigtable/pull/1571/files#r2114536705)

Neither of those need to block this, so we can open issues to pick them up later if needed

danieljbruce committed:
Add a check to make sure options exist

mutianf reviewed (dismissed):
LGTM after the final nits

danieljbruce committed:
Do console warn. Not throw error

danieljbruce committed:
Debugging attempts

danieljbruce committed:
Remove the sinon clock

danieljbruce committed:
Add all the tests back in

danieljbruce committed:
still include the sinon stub

danieljbruce committed:
Remove chunk transformer handler

danieljbruce committed:
formatting changes

danieljbruce committed:
Add space

danieljbruce committed:
Add space

danieljbruce committed:
remove only

danieljbruce committed:
Merge branch 'just-pick-failing-readrow-test' of https://github.com/googleapis/nodejs-bigtable into 359913994-third-PR-CSM

# Conflicts:
#	system-test/read-rows.ts

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into 359913994-third-PR-CSM

# Conflicts:
#	system-test/read-rows.ts
